contrib_bots/lib/giveaway.py

```diff
@@ -0,0 +1,211 @@
+    to run for. Full syntax is shown below:
+    @giveaway command giveaway_name giveaway_time
+
+    Possible commands:
+    init - initialize a giveaway (only one can be running at a time)
+    enter - enter a giveaway (a giveaway must be running to use this command)
+    exit - leave a giveaway (must be partaking in an active giveaway for this
+    command to work
+    cancel - cancel a giveaway (can only be run by the person who first
+    initialised the giveaway
+
+    This plugin was created by Daniel O'Brien
+    as part of Google Code-In 2016-2017
+    '''
+    def __init__(self):
+        self.giveawayArray = []
+        self.giveawayOn = False
+        self.giveawayName = ""
+        self.giveawayTime = 0
+        self.giveawayCreator = ""
+        self.giveawayStartTime = 0
+        self.giveawayGroup = ""
+        self.giveawaySubject = ""
+        self.giveawayClient = ""
+        self.giveawayTimeMins = 0
+        self.giveawayStage = 0
+        self.one_minute_warning = False
+        self.giveawayWinner = ""
+        self.winnerContactDetails = []
+        self.winnerIndex = 0
+
+    def time_update(self, new_content):
+            if self.giveawayOn == True:
+                self.giveawayClient.send_message(dict(
+                    type='stream',
+                    to=self.giveawayStream,
+                    subject=self.giveawaySubject,
+                    content=new_content,
+                ))
+                threading.Timer(6.0, self.check_time).start()
+
+    def check_time(self):
+        if time.time() > (self.giveawayStartTime+self.giveawayTime):
+            print("Giveaway ended")
+            try:
+                self.giveawayWinner = random.choice(self.giveawayArray)
+                self.winnerIndex = self.giveawayArray.index(self.giveawayWinner)
+                new_content = 'The giveaway has now ended... The winner is... %s!' % (self.giveawayWinner)
+                direct_content = """Congratulations!, you recently partook in the giveaway called %s and won!
+                Please contact %s on details as to how to claim your prize.""" % (self.giveawayName, self.giveawayCreator,)
+                self.giveawayClient.send_message(dict(
+                    type='private',
+                    to=self.winnerContactDetails[self.winnerIndex],
+                    content=direct_content,
+                ))
+
+            except:
+                new_content = 'Giveaway ended: No one entered the giveaway :('
+            if self.giveawayOn == True:
+                self.giveawayOn = False
+                self.giveawayClient.send_message(dict(
+                    type='stream',
+                    to=self.giveawayStream,
+                    subject=self.giveawaySubject,
+                    content=new_content,
+                ))
+        elif time.time() > (self.giveawayStartTime+(old_div(self.giveawayTime, 4))) and self.giveawayStage < 1:
```

Comment:
@TomaszKolek went one step further and have managed to compress all of the if statements into a single if statement